src/CustomElements/v1/CustomElements.js

```diff
@@ -0,0 +1,597 @@
+    // the fourth argument be present even though the other browsers ignore it.
+    return doc.createTreeWalker(root, NodeFilter.SHOW_ELEMENT, null, false);
+  }
+
+  function isElement(node) {
+    return node.nodeType === Node.ELEMENT_NODE
+  }
+
+  /**
+   * A registry of custom element definitions.
+   *
+   * See https://html.spec.whatwg.org/multipage/scripting.html#customelementsregistry
+   *
+   * @constructor
+   * @property {boolean} polyfilled Whether this registry is polyfilled
+   * @property {boolean} enableFlush Set to true to enable the flush() method
+   *   to work. This should only be done for tests, as it causes a memory leak.
+   */
+  function CustomElementsRegistry() {
+    /** @private {Map<string, CustomElementDefinition>} **/
+    this._definitions = new Map();
+
+    /** @private {Map<Function, CustomElementDefinition>} **/
+    this._constructors = new Map();
+
+    this._whenDefinedMap = new Map();
+
+    /** @private {Set<MutationObserver>} **/
+    this._observers = new Set();
+
+    /** @private {MutationObserver} **/
+    this._attributeObserver =
+        new MutationObserver(this._handleAttributeChange.bind(this));
+
+    /** @private {HTMLElement} **/
+    this._newInstance = null;
+
+    this.polyfilled = true;
+    this.enableFlush = false;
+
+    this._observeRoot(document);
+  }
+
+  CustomElementsRegistry.prototype = {
+
+    // HTML spec part 4.13.4
+    // https://html.spec.whatwg.org/multipage/scripting.html#dom-customelementsregistry-define
+    define: function(name, constructor, options) {
+      name = name.toString().toLowerCase();
+
+      // 1:
+      if (typeof constructor !== 'function') {
+        throw new TypeError('constructor must be a Constructor');
+      }
+
+      // 2. If constructor is an interface object whose corresponding interface
+      //    either is HTMLElement or has HTMLElement in its set of inherited
+      //    interfaces, throw a TypeError and abort these steps.
+      //
+      // It doesn't appear possible to check this condition from script
+
+      // 3:
+      if (!isValidCustomElementName(name)) {
+        throw new SyntaxError(`The element name '${name}' is not valid.`);
+      }
+
+      // 4, 5:
+      // Note: we don't track being-defined names and constructors because
+      // define() isn't normally reentrant. The only time user code can run
+      // during define() is when getting callbacks off the prototype, which
+      // would be highly-unusual. We can make define() reentrant-safe if needed.
+      if (this._definitions.has(name)) {
+        throw new Error(`An element with name '${name}' is already defined`);
+      }
+
+      // 6, 7:
+      if (this._constructors.has(constructor)) {
+        throw new Error(`Definition failed for '${name}': ` +
+            `The constructor is already used.`);
+      }
+
+      // 8:
+      var localName = name;
+
+      // 9, 10: We do not support extends currently.
```

Comment:
Is type extension planned? If so, maybe we can add a note somewhere that it's a planned feature?